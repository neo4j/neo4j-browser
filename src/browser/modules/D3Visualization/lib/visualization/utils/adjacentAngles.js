/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

export default class AdjacentAngles {
  findRuns(AngleList, minSeparation) {
    let p = 0
    let start = 0
    let end = 0
    const runs = []
    const minStart = function() {
      if (runs.length === 0) {
        return 0
      } else {
        return runs[0].start
      }
    }

    const scanForDensePair = function() {
      start = p
      end = AngleList.wrapIndex(p + 1)
      if (end === minStart()) {
        return 'done'
      } else {
        p = end
        if (tooDense(start, end)) {
          return extendEnd
        } else {
          return scanForDensePair
        }
      }
    }

    const extendEnd = function() {
      if (p === minStart()) {
        return 'done'
      } else if (tooDense(start, AngleList.wrapIndex(p + 1))) {
        end = AngleList.wrapIndex(p + 1)
        p = end
        return extendEnd
      } else {
        p = start
        return extendStart
      }
    }

    const extendStart = function() {
      const candidateStart = AngleList.wrapIndex(p - 1)
      if (tooDense(candidateStart, end) && candidateStart !== end) {
        start = candidateStart
        p = start
        return extendStart
      } else {
        runs.push({
          start,
          end
        })
        p = end
        return scanForDensePair
      }
    }

    const tooDense = function(start, end) {
      const run = {
        start,
        end
      }
      return AngleList.angle(run) < AngleList.length(run) * minSeparation
    }

    let stepCount = 0
    let step = scanForDensePair
    while (step !== 'done') {
      if (stepCount++ > AngleList.totalLength() * 10) {
        console.log(
          'Warning: failed to layout arrows',
          (() => {
            const result = []
            for (const key of Object.keys(AngleList.list || {})) {
              const value = AngleList.list[key]
              result.push(`${key}: ${value.angle}`)
            }
            return result
          })().join('\n'),
          minSeparation
        )
        break
      }
      step = step()
    }

    return runs
  }
}
