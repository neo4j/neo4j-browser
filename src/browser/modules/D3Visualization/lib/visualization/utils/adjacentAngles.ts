/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import AngleList from './angleList'

export type Run = { start: number; end: number }

export default class AdjacentAngles {
  findRuns(angleList: AngleList, minSeparation: number): Run[] {
    let p = 0
    let start = 0
    let end = 0
    const runs: Run[] = []
    const minStart = function () {
      if (runs.length === 0) {
        return 0
      } else {
        return runs[0].start
      }
    }

    const scanForDensePair = function () {
      start = p
      end = angleList.wrapIndex(p + 1)
      if (end === minStart()) {
        return 'done'
      } else {
        p = end
        if (tooDense(start, end)) {
          return extendEnd
        } else {
          return scanForDensePair
        }
      }
    }

    const extendEnd = function () {
      if (p === minStart()) {
        return 'done'
      } else if (tooDense(start, angleList.wrapIndex(p + 1))) {
        end = angleList.wrapIndex(p + 1)
        p = end
        return extendEnd
      } else {
        p = start
        return extendStart
      }
    }

    const extendStart = function () {
      const candidateStart = angleList.wrapIndex(p - 1)
      if (tooDense(candidateStart, end) && candidateStart !== end) {
        start = candidateStart
        p = start
        return extendStart
      } else {
        runs.push({
          start,
          end
        })
        p = end
        return scanForDensePair
      }
    }

    const tooDense = function (start: number, end: number) {
      const run = {
        start,
        end
      }
      return angleList.angle(run) < angleList.length(run) * minSeparation
    }

    // Recursive function that's hard to properly type
    let step: 'done' | (() => any) = scanForDensePair

    let stepCount = 0
    while (step !== 'done') {
      if (stepCount++ > angleList.totalLength() * 10) {
        console.log(
          'Warning: failed to layout arrows',
          Object.entries(angleList)
            .map(([key, value]) => `${key}: ${value.angle}`)
            .join('\n'),
          minSeparation
        )
        break
      }
      step = step()
    }

    return runs
  }
}
