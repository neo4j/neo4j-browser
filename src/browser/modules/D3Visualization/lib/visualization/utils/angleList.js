/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

export default class AngleList {
  constructor(list) {
    this.list = list
  }

  getAngle(index) {
    return this.list[index].angle
  }

  fixed(index) {
    return this.list[index].fixed
  }

  totalLength() {
    return this.list.length
  }

  length(run) {
    if (run.start < run.end) {
      return run.end - run.start
    } else {
      return run.end + this.list.length - run.start
    }
  }

  angle(run) {
    if (run.start < run.end) {
      return this.list[run.end].angle - this.list[run.start].angle
    } else {
      return 360 - (this.list[run.start].angle - this.list[run.end].angle)
    }
  }

  wrapIndex(index) {
    if (index === -1) {
      return this.list.length - 1
    } else if (index >= this.list.length) {
      return index - this.list.length
    } else {
      return index
    }
  }
}
