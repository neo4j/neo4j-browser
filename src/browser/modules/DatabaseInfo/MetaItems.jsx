/*
 * Copyright (c) 2002-2017 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import classNames from 'classnames'
import styles from './style_meta.css'
import {DrawerSubHeader, DrawerSection, DrawerSectionBody} from 'browser-components/drawer'
import {StyledLabel, StyledRelationship, StyledProperty} from './styled'

const createItems = (originalList, onItemClick, RenderType, editorCommandTemplate, showStar = true) => {
  let items = [...originalList]
  if (showStar) {
    items.unshift('*')
  }
  return items.map((text, index) => {
    const getNodesCypher = editorCommandTemplate(text)
    return (
      <RenderType.component
        key={index}
        onClick={() => onItemClick(getNodesCypher)}>
        {text}
      </RenderType.component>
    )
  })
}
const LabelItems = ({labels, onItemClick}) => {
  let labelItems = <p>There are no labels in database</p>
  if (labels.length > 0) {
    const editorCommandTemplate = (text) => {
      if (text === '*') {
        return 'MATCH (n) RETURN n LIMIT 25'
      }
      return `MATCH (n:${text}) RETURN n LIMIT 25`
    }
    labelItems = createItems(labels, onItemClick, {component: StyledLabel}, editorCommandTemplate)
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Node Labels</DrawerSubHeader>
      <DrawerSectionBody className={classNames({
        [styles['wrapper']]: true
      })}>
        {labelItems}
      </DrawerSectionBody>
    </DrawerSection>
  )
}
const RelationshipItems = ({relationshipTypes, onItemClick}) => {
  let relationshipItems = <p>No relationships in database</p>
  if (relationshipTypes.length > 0) {
    const editorCommandTemplate = (text) => {
      if (text === '*') {
        return 'MATCH p=()-->() RETURN p LIMIT 25'
      }
      return `MATCH p=()-[r:${text}]->() RETURN p LIMIT 25`
    }
    relationshipItems = createItems(relationshipTypes, onItemClick, {component: StyledRelationship}, editorCommandTemplate)
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Relationship Types</DrawerSubHeader>
      <DrawerSectionBody className={classNames({
        [styles['wrapper']]: true
      })}>
        {relationshipItems}
      </DrawerSectionBody>
    </DrawerSection>
  )
}
const PropertyItems = ({properties, onItemClick}) => {
  let propertyItems = <p>There are no properites in database</p>
  if (properties.length > 0) {
    const editorCommandTemplate = (text) => {
      return `MATCH (n) WHERE EXISTS(n.${text}) RETURN DISTINCT "node" as entity, n.${text} AS ${text} LIMIT 25 UNION ALL MATCH ()-[r]-() WHERE EXISTS(r.${text}) RETURN DISTINCT "relationship" AS entity, r.${text} AS ${text} LIMIT 25`
    }
    propertyItems = createItems(properties, onItemClick, {component: StyledProperty}, editorCommandTemplate, false)
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Property Keys</DrawerSubHeader>
      <DrawerSectionBody className={classNames({
        [styles['wrapper']]: true
      })}>
        {propertyItems}
      </DrawerSectionBody>
    </DrawerSection>
  )
}

export {
  LabelItems,
  RelationshipItems,
  PropertyItems
}
