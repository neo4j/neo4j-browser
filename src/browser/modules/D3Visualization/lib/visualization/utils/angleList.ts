/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { Run } from './adjacentAngles'

type Angle = {
  key: string
  angle: number
  fixed: boolean
}

export default class AngleList {
  list: Angle[]
  constructor(list: Angle[]) {
    this.list = list
  }

  getAngle(index: number): number {
    return this.list[index].angle
  }

  fixed(index: number): boolean | undefined {
    return this.list[index].fixed
  }

  totalLength(): number {
    return this.list.length
  }

  length(run: Run): number {
    if (run.start < run.end) {
      return run.end - run.start
    } else {
      return run.end + this.list.length - run.start
    }
  }

  angle(run: Run): number {
    if (run.start < run.end) {
      return this.list[run.end].angle - this.list[run.start].angle
    } else {
      return 360 - (this.list[run.start].angle - this.list[run.end].angle)
    }
  }

  wrapIndex(index: number): number {
    if (index === -1) {
      return this.list.length - 1
    } else if (index >= this.list.length) {
      return index - this.list.length
    } else {
      return index
    }
  }
}
