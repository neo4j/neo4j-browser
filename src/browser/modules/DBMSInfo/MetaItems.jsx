/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import { ecsapeCypherMetaItem } from 'services/utils'
import classNames from 'classnames'
import styles from './style_meta.css'
import {
  DrawerSubHeader,
  DrawerSection,
  DrawerSectionBody
} from 'browser-components/drawer'
import {
  StyledLabel,
  StyledRelationship,
  StyledProperty,
  StyledShowMoreContainer,
  StyledShowMoreLink
} from './styled'
import Render from 'browser-components/Render'
import numberToUSLocale from 'shared/utils/number-to-US-locale'

const ShowMore = ({ total, shown, moreStep, onMore }) => {
  const numMore = total - shown > moreStep ? moreStep : total - shown
  return (
    <Render if={shown < total}>
      <StyledShowMoreContainer>
        <StyledShowMoreLink onClick={() => onMore(numMore)}>
          Show {numMore} more
        </StyledShowMoreLink>
        &nbsp;|&nbsp;
        <StyledShowMoreLink onClick={() => onMore(total)}>
          Show all
        </StyledShowMoreLink>
      </StyledShowMoreContainer>
    </Render>
  )
}

const createItems = (
  originalList,
  onItemClick,
  RenderType,
  editorCommandTemplate,
  showStar = true,
  count
) => {
  let items = [...originalList]
  if (showStar) {
    let str = '*'
    if (count) {
      str = `${str}(${numberToUSLocale(count)})`
    }
    items.unshift(str)
  }
  return items.map((text, index) => {
    const getNodesCypher = editorCommandTemplate(text, index)
    return (
      <RenderType.component
        data-testid='sidebarMetaItem'
        key={index}
        onClick={() => onItemClick(getNodesCypher)}
      >
        {text}
      </RenderType.component>
    )
  })
}
const LabelItems = ({
  labels = [],
  totalNumItems,
  onItemClick,
  moreStep,
  onMoreClick,
  count
}) => {
  let labelItems = <p>There are no labels in database</p>
  if (labels.length) {
    const editorCommandTemplate = (text, i) => {
      if (i === 0) {
        return 'MATCH (n) RETURN n LIMIT 25'
      }
      return `MATCH (n:${ecsapeCypherMetaItem(text)}) RETURN n LIMIT 25`
    }
    labelItems = createItems(
      labels,
      onItemClick,
      { component: StyledLabel },
      editorCommandTemplate,
      true,
      count
    )
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Node Labels</DrawerSubHeader>
      <DrawerSectionBody
        className={classNames({
          [styles['wrapper']]: true
        })}
      >
        {labelItems}
      </DrawerSectionBody>
      <ShowMore
        total={totalNumItems}
        shown={labels.length}
        moreStep={moreStep}
        onMore={onMoreClick}
      />
    </DrawerSection>
  )
}
const RelationshipItems = ({
  relationshipTypes = [],
  totalNumItems,
  onItemClick,
  moreStep,
  onMoreClick,
  count
}) => {
  let relationshipItems = <p>No relationships in database</p>
  if (relationshipTypes.length > 0) {
    const editorCommandTemplate = (text, i) => {
      if (i === 0) {
        return 'MATCH p=()-->() RETURN p LIMIT 25'
      }
      return `MATCH p=()-[r:${ecsapeCypherMetaItem(
        text
      )}]->() RETURN p LIMIT 25`
    }
    relationshipItems = createItems(
      relationshipTypes,
      onItemClick,
      { component: StyledRelationship },
      editorCommandTemplate,
      true,
      count
    )
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Relationship Types</DrawerSubHeader>
      <DrawerSectionBody
        className={classNames({
          [styles['wrapper']]: true
        })}
      >
        {relationshipItems}
      </DrawerSectionBody>
      <ShowMore
        total={totalNumItems}
        shown={relationshipTypes.length}
        moreStep={moreStep}
        onMore={onMoreClick}
      />
    </DrawerSection>
  )
}
const PropertyItems = ({
  properties,
  totalNumItems,
  onItemClick,
  moreStep,
  onMoreClick
}) => {
  let propertyItems = <p>There are no properties in database</p>
  if (properties.length > 0) {
    const editorCommandTemplate = text => {
      return `MATCH (n) WHERE EXISTS(n.${ecsapeCypherMetaItem(
        text
      )}) RETURN DISTINCT "node" as entity, n.${ecsapeCypherMetaItem(
        text
      )} AS ${ecsapeCypherMetaItem(
        text
      )} LIMIT 25 UNION ALL MATCH ()-[r]-() WHERE EXISTS(r.${ecsapeCypherMetaItem(
        text
      )}) RETURN DISTINCT "relationship" AS entity, r.${ecsapeCypherMetaItem(
        text
      )} AS ${ecsapeCypherMetaItem(text)} LIMIT 25`
    }
    propertyItems = createItems(
      properties,
      onItemClick,
      { component: StyledProperty },
      editorCommandTemplate,
      false
    )
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Property Keys</DrawerSubHeader>
      <DrawerSectionBody
        className={classNames({
          [styles['wrapper']]: true
        })}
      >
        {propertyItems}
      </DrawerSectionBody>
      <ShowMore
        total={totalNumItems}
        shown={properties.length}
        moreStep={moreStep}
        onMore={onMoreClick}
      />
    </DrawerSection>
  )
}

export { LabelItems, RelationshipItems, PropertyItems }
