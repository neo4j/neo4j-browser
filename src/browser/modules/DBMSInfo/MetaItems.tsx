/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import deepmerge from 'deepmerge'
import React from 'react'
import { ThemeProvider } from 'styled-components'

import { GraphStyleModel } from 'graph-visualization'

import styles from './style_meta.css'
import { StyledLabel, StyledProperty, StyledRelationship } from './styled'
import { ShowMoreOrAll } from 'browser-components/ShowMoreOrAll/ShowMoreOrAll'
import {
  DrawerSection,
  DrawerSectionBody,
  DrawerSubHeader
} from 'browser-components/drawer/drawer-styled'
import { dark } from 'browser-styles/themes'
import { escapeCypherIdentifier } from 'services/utils'
import numberToUSLocale from 'shared/utils/number-to-US-locale'

const wrapperStyle = (styles && styles.wrapper) || ''

function createStyleGetter(graphStyleData: any, kind: string) {
  const graphStyle = new GraphStyleModel()
  if (graphStyleData) {
    graphStyle.loadRules(deepmerge(graphStyle.toSheet(), graphStyleData || {}))
  }

  if (kind === 'node') {
    return function (text: string) {
      if (graphStyleData) {
        const styleForItem = graphStyle.forNode({
          labels: [text]
        })
        return {
          backgroundColor: styleForItem.get('color'),
          color: styleForItem.get('text-color-internal')
        }
      }
      return {}
    }
  }
  if (kind === 'relationship') {
    return function (text: string) {
      if (graphStyleData) {
        const styleForItem = graphStyle.forRelationship({
          type: text
        })
        return {
          backgroundColor: styleForItem.get('color'),
          color: styleForItem.get('text-color-internal')
        }
      }
      return {}
    }
  }
  throw new Error(`Unsupported argument: ${kind}`)
}
function createNodeStyleGetter(graphStyleData: any) {
  return createStyleGetter(graphStyleData, 'node')
}
function createRelationshipStyleGetter(graphStyleData: any) {
  return createStyleGetter(graphStyleData, 'relationship')
}

const createItems = (
  originalList: any,
  onItemClick: any,
  RenderType: any,
  editorCommandTemplate: any,
  showStar = true,
  count?: number,
  styleGetter: any = () => ({})
) => {
  const items = [...originalList]
  if (showStar) {
    let str = '*'
    if (count) {
      str = `${str}(${numberToUSLocale(count)})`
    }
    items.unshift(str)
  }

  return items.map((text, index) => {
    const getNodesCypher = editorCommandTemplate(text, index)
    return (
      <RenderType.component
        data-testid="sidebarMetaItem"
        key={index}
        onClick={() => onItemClick(getNodesCypher)}
        style={styleGetter(text)}
      >
        {text}
      </RenderType.component>
    )
  })
}

type LabelItemsProps = {
  labels: string[]
  totalNumItems: number
  onItemClick: () => void
  moreStep: number
  onMoreClick: (num: number) => void
  count: number
  graphStyleData: any
}
const LabelItems = ({
  labels = [],
  totalNumItems,
  onItemClick,
  moreStep,
  onMoreClick,
  count,
  graphStyleData
}: LabelItemsProps) => {
  let labelItems: any = <p>There are no labels in database</p>
  if (labels.length) {
    const editorCommandTemplate = (text: any, i: any) => {
      if (i === 0) {
        return 'MATCH (n) RETURN n LIMIT 25'
      }
      return `MATCH (n:${escapeCypherIdentifier(text)}) RETURN n LIMIT 25`
    }
    labelItems = createItems(
      labels,
      onItemClick,
      { component: StyledLabel },
      editorCommandTemplate,
      true,
      count,
      createNodeStyleGetter(graphStyleData)
    )
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Node Labels</DrawerSubHeader>
      <DrawerSectionBody className={wrapperStyle}>
        {labelItems}
      </DrawerSectionBody>
      <ThemeProvider theme={dark}>
        <ShowMoreOrAll
          total={totalNumItems}
          shown={labels.length}
          moreStep={moreStep}
          onMore={onMoreClick}
        />
      </ThemeProvider>
    </DrawerSection>
  )
}

type RelationshipItemsProps = {
  relationshipTypes: string[]
  totalNumItems: number
  onItemClick: () => void
  moreStep: number
  onMoreClick: (num: number) => any
  count: number
  graphStyleData: any
}
const RelationshipItems = ({
  relationshipTypes = [],
  totalNumItems,
  onItemClick,
  moreStep,
  onMoreClick,
  count,
  graphStyleData
}: RelationshipItemsProps) => {
  let relationshipItems: any = <p>No relationships in database</p>
  if (relationshipTypes.length > 0) {
    const editorCommandTemplate = (text: any, i: any) => {
      if (i === 0) {
        return 'MATCH p=()-->() RETURN p LIMIT 25'
      }
      return `MATCH p=()-[r:${escapeCypherIdentifier(
        text
      )}]->() RETURN p LIMIT 25`
    }
    relationshipItems = createItems(
      relationshipTypes,
      onItemClick,
      { component: StyledRelationship },
      editorCommandTemplate,
      true,
      count,
      createRelationshipStyleGetter(graphStyleData)
    )
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Relationship Types</DrawerSubHeader>
      <DrawerSectionBody className={wrapperStyle}>
        {relationshipItems}
      </DrawerSectionBody>
      <ShowMoreOrAll
        total={totalNumItems}
        shown={relationshipTypes.length}
        moreStep={moreStep}
        onMore={onMoreClick}
      />
    </DrawerSection>
  )
}

type PropertyItemsProps = {
  properties: string[]
  totalNumItems: number
  onItemClick: () => void
  moreStep: number
  onMoreClick: (num: number) => any
}
const PropertyItems = ({
  properties,
  totalNumItems,
  onItemClick,
  moreStep,
  onMoreClick
}: PropertyItemsProps) => {
  let propertyItems: any = <p>There are no properties in database</p>
  if (properties.length > 0) {
    const editorCommandTemplate = (text: any) => {
      return `MATCH (n) WHERE EXISTS(n.${escapeCypherIdentifier(
        text
      )}) RETURN DISTINCT "node" as entity, n.${escapeCypherIdentifier(
        text
      )} AS ${escapeCypherIdentifier(
        text
      )} LIMIT 25 UNION ALL MATCH ()-[r]-() WHERE EXISTS(r.${escapeCypherIdentifier(
        text
      )}) RETURN DISTINCT "relationship" AS entity, r.${escapeCypherIdentifier(
        text
      )} AS ${escapeCypherIdentifier(text)} LIMIT 25`
    }
    propertyItems = createItems(
      properties,
      onItemClick,
      { component: StyledProperty },
      editorCommandTemplate,
      false
    )
  }
  return (
    <DrawerSection>
      <DrawerSubHeader>Property Keys</DrawerSubHeader>
      <DrawerSectionBody className={wrapperStyle}>
        {propertyItems}
      </DrawerSectionBody>
      <ShowMoreOrAll
        total={totalNumItems}
        shown={properties.length}
        moreStep={moreStep}
        onMore={onMoreClick}
      />
    </DrawerSection>
  )
}

export { LabelItems, RelationshipItems, PropertyItems }
