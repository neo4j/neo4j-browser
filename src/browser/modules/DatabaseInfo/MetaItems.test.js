/*
 * Copyright (c) 2002-2018 "Neo4j, Inc"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* global jest, test, expect */

import { mount } from 'services/testUtils'
import { ecsapeCypherMetaItem } from 'services/utils'
import { LabelItems, RelationshipItems } from './MetaItems.jsx'

test('LabelItems detects user defined label named * and creates query correctly', () => {
  // Given
  const onItemClick = jest.fn()
  const labels = ['*', 'Label1']

  // When
  return mount(LabelItems)
    .withProps({
      labels,
      totalNumItems: labels.length,
      onItemClick
    })
    .then(wrapper => {
      const items = wrapper.find(`[data-test-id='sidebarMetaItem']`)
      items.forEach((item, i) => {
        const element = item.get(0)
        element.click()
        const expectedCall =
          i === 0
            ? 'MATCH (n) RETURN n LIMIT 25'
            : `MATCH (n:${ecsapeCypherMetaItem(item.text())}) RETURN n LIMIT 25`
        expect(onItemClick).toHaveBeenLastCalledWith(expectedCall)
      })
    })
})

test('RelationshipItems detects user defined relType named * and creates query correctly', () => {
  // Given
  const onItemClick = jest.fn()
  const relationshipTypes = ['*', 'TYPE1']

  // When
  return mount(RelationshipItems)
    .withProps({
      relationshipTypes,
      totalNumItems: relationshipTypes.length,
      onItemClick
    })
    .then(wrapper => {
      const items = wrapper.find(`[data-test-id='sidebarMetaItem']`)
      items.forEach((item, i) => {
        const element = item.get(0)
        element.click()
        const expectedCall =
          i === 0
            ? 'MATCH p=()-->() RETURN p LIMIT 25'
            : `MATCH p=()-[r:${ecsapeCypherMetaItem(
              item.text()
            )}]->() RETURN p LIMIT 25`
        expect(onItemClick).toHaveBeenLastCalledWith(expectedCall)
      })
    })
})
