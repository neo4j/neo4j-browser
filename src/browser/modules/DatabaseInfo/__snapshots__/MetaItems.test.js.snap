// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelItems renders empty 1`] = `
<div>
  <div
    class="Drawer__DrawerSection-sc-5bct1g-4 bEhGCY"
  >
    <h5
      class="Drawer__DrawerSubHeader-sc-5bct1g-3 laKfsi"
    >
      Node Labels
    </h5>
    <div
      class="undefined Drawer__DrawerSectionBody-sc-5bct1g-5 RMcRs"
    >
      <p>
        There are no labels in database
      </p>
    </div>
  </div>
</div>
`;

exports[`LabelItems renders labels 1`] = `
<div>
  <div
    class="Drawer__DrawerSection-sc-5bct1g-4 bEhGCY"
  >
    <h5
      class="Drawer__DrawerSubHeader-sc-5bct1g-3 laKfsi"
    >
      Node Labels
    </h5>
    <div
      class="undefined Drawer__DrawerSectionBody-sc-5bct1g-5 RMcRs"
    >
      <div
        class="styled__chip-sc-1srdf8s-0 styled__StyledLabel-sc-1srdf8s-1 cnkXFe"
        data-testid="sidebarMetaItem"
      >
        *
      </div>
      <div
        class="styled__chip-sc-1srdf8s-0 styled__StyledLabel-sc-1srdf8s-1 cnkXFe"
        data-testid="sidebarMetaItem"
      >
        MyLabel
      </div>
      <div
        class="styled__chip-sc-1srdf8s-0 styled__StyledLabel-sc-1srdf8s-1 cnkXFe"
        data-testid="sidebarMetaItem"
      >
        MyLabel2
      </div>
    </div>
  </div>
</div>
`;

exports[`PropertyItems renders empty 1`] = `
<div>
  <div
    class="Drawer__DrawerSection-sc-5bct1g-4 bEhGCY"
  >
    <h5
      class="Drawer__DrawerSubHeader-sc-5bct1g-3 laKfsi"
    >
      Property Keys
    </h5>
    <div
      class="undefined Drawer__DrawerSectionBody-sc-5bct1g-5 RMcRs"
    >
      <p>
        There are no properties in database
      </p>
    </div>
  </div>
</div>
`;

exports[`PropertyItems renders properties 1`] = `
<div>
  <div
    class="Drawer__DrawerSection-sc-5bct1g-4 bEhGCY"
  >
    <h5
      class="Drawer__DrawerSubHeader-sc-5bct1g-3 laKfsi"
    >
      Property Keys
    </h5>
    <div
      class="undefined Drawer__DrawerSectionBody-sc-5bct1g-5 RMcRs"
    >
      <div
        class="styled__chip-sc-1srdf8s-0 styled__StyledProperty-sc-1srdf8s-3 EddIa"
        data-testid="sidebarMetaItem"
      >
        prop1
      </div>
      <div
        class="styled__chip-sc-1srdf8s-0 styled__StyledProperty-sc-1srdf8s-3 EddIa"
        data-testid="sidebarMetaItem"
      >
        prop2
      </div>
    </div>
  </div>
</div>
`;

exports[`RelationshipItems renders empty 1`] = `
<div>
  <div
    class="Drawer__DrawerSection-sc-5bct1g-4 bEhGCY"
  >
    <h5
      class="Drawer__DrawerSubHeader-sc-5bct1g-3 laKfsi"
    >
      Relationship Types
    </h5>
    <div
      class="undefined Drawer__DrawerSectionBody-sc-5bct1g-5 RMcRs"
    >
      <p>
        No relationships in database
      </p>
    </div>
  </div>
</div>
`;

exports[`RelationshipItems renders relationshipTypes 1`] = `
<div>
  <div
    class="Drawer__DrawerSection-sc-5bct1g-4 bEhGCY"
  >
    <h5
      class="Drawer__DrawerSubHeader-sc-5bct1g-3 laKfsi"
    >
      Relationship Types
    </h5>
    <div
      class="undefined Drawer__DrawerSectionBody-sc-5bct1g-5 RMcRs"
    >
      <div
        class="styled__chip-sc-1srdf8s-0 styled__StyledRelationship-sc-1srdf8s-2 fcKIry"
        data-testid="sidebarMetaItem"
      >
        *
      </div>
      <div
        class="styled__chip-sc-1srdf8s-0 styled__StyledRelationship-sc-1srdf8s-2 fcKIry"
        data-testid="sidebarMetaItem"
      >
        MY_TYPE
      </div>
      <div
        class="styled__chip-sc-1srdf8s-0 styled__StyledRelationship-sc-1srdf8s-2 fcKIry"
        data-testid="sidebarMetaItem"
      >
        MY_TYPE2
      </div>
    </div>
  </div>
</div>
`;
